symbolic-symcache/src/new/writer.rs

```diff
@@ -308,6 +313,113 @@ impl SymCacheConverter {
         }
     }
 
+    #[cfg(feature = "il2cpp")]
+    /// Processes a set of [`UsymSymbols`], passing all mapped symbols into the converter.
+    pub fn process_usym(&mut self, usym: &UsymSymbols) -> Result<(), SymCacheError> {
+        // Assume records they are sorted by address; There's a test that guarantees this
+        let mapped_records = usym.records().filter_map(|r| match r {
+            UsymSourceRecord::Unmapped(_) => None,
+            UsymSourceRecord::Mapped(r) => Some(r),
+        });
+
+        let mut curr_id: String = String::default();
+        let mut function_idx = 0;
+        for record in mapped_records {
+            // like process_symbolic_function, skip functions whose address is too large to fit in a
+            // u32
+            if record.address > u32::MAX as u64 {
+                continue;
+            }
+            let address = record.address as u32;
+
+            // Records that belong to the same function will have the same identifier.
+            // Symbols have GUID-like sections to them that might ensure they're unique across
+            // files, but we'll just include the file name and paths to be very safe.
+            let identifier = format!("{}::{}", record.native_file, record.native_symbol);
+            if identifier != curr_id {
+                function_idx = {
+                    let mut function = transform::Function {
+                        name: record.managed_symbol.clone(),
+                        comp_dir: None,
+                    };
+                    for transformer in &self.transformers.0 {
+                        function = transformer.transform_function(function);
+                    }
+
+                    let string_bytes = &mut self.string_bytes;
+                    let strings = &mut self.strings;
+                    let name_offset = Self::insert_string(string_bytes, strings, &function.name);
+
+                    let (fun_idx, _) = self.functions.insert_full(raw::Function {
```

Comment:
I explored the idea of changing `raw::Function`'s `Hash` and `PartialEq` implementations so they're only considered equivalent if the `name_offset` is the same. Turns out this breaks a bunch of existing symcache tests by omitting a bunch of different functions, so I gave up on that idea.

For posterity, this is an example of some of the functions that go missing after implementing this idea:



```
Snapshot file: symbolic-symcache/tests/snapshots/test_writer__functions_macos.snap
Snapshot: functions_macos
Source: symbolic-symcache/tests/test_writer.rs:127

   84    72 │             7410 _ZN15google_breakpad17MinidumpGenerator10WriteStackEPjP18MDMemoryDescriptor
   85       │-            7440 _ZN15google_breakpad17MinidumpGenerator13WriteStackX86EPjP18MDMemoryDescriptor
   86       │-            7450 _ZN15google_breakpad17MinidumpGenerator16WriteStackX86_64EPjP18MDMemoryDescriptor
   87    73 │             7460 _ZN15google_breakpad17MinidumpGenerator12WriteContextEPjP20MDLocationDescriptor
   88    74 │             7490 _ZN15google_breakpad17MinidumpGenerator15WriteContextX86EPjP20MDLocationDescriptor
   89    75 │             7640 _ZN15google_breakpad17MinidumpGenerator18WriteContextX86_64EPjP20MDLocationDescriptor
   90       │-            7810 _ZN15google_breakpad17MinidumpGenerator17CurrentPCForStackEPj
   91       │-            7850 _ZN15google_breakpad17MinidumpGenerator20CurrentPCForStackX86EPj
   92       │-            7860 _ZN15google_breakpad17MinidumpGenerator23CurrentPCForStackX86_64EPj
   93       │-            7870 _ZN15google_breakpad17MinidumpGenerator14GetThreadStateEjPjS1_
   94    76 │             7930 _ZN15google_breakpad17MinidumpGenerator17WriteThreadStreamEjP11MDRawThread
   95    77 │             7ae0 _ZN15google_breakpad17MinidumpGenerator20WriteExceptionStreamEP14MDRawDirectory
   96    78 │             7d60 _ZN15google_breakpad17MinidumpGenerator17WriteModuleStreamEjP11MDRawModule
   97       │-            8100 _ZN15google_breakpad17MinidumpGenerator20FindExecutableModuleEv
   98    79 │             8150 _ZN15google_breakpad17MinidumpGenerator13WriteCVRecordEP11MDRawModuleiPKcb
   99    80 │             8500 _ZN15google_breakpad13DynamicImagesD2Ev
  100       │-            85b0 _ZNSt3__114__split_bufferI18MDMemoryDescriptorRN15google_breakpad16PageStdAllocatorIS1_EEEC2EmmS5_
  101    81 │             8710 _ZNSt3__16vectorI18MDMemoryDescriptorN15google_breakpad16PageStdAllocatorIS1_EEE21__push_back_slow_pathIRKS1_EEvOT_
```

Reply:
cc @loewenheim

Reply:
Ugh, I think I know why this is.

1. In a symbolic object, there can be multiple functions with the same name because every inlined copy is represented by a distinct function.
2. The first time the writer encounters the missing function, it's an inlined version. It gets inserted into the writer with an `entry_pc` of `u32::MAX`: https://github.com/getsentry/symbolic/blob/047cb033695eaa844fd7bf43af473b1c74e25376/symbolic-symcache/src/new/writer.rs#L144)
3. When the non-inlined version of the function is encountered later, it doesn't get inserted again because a function that compares as equal is already there.
4. `FunctionsDebug`, which is used to generate these snapshots, filters out functions with an `address` (= `entry_pc`) of `u32::MAX` so it doesn't uselessly display every inlined copy. But now we have no more copies, we only have the inlined one. https://github.com/getsentry/symbolic/blob/047cb033695eaa844fd7bf43af473b1c74e25376/symbolic-symcache/tests/test_writer.rs#L24

Reply:
i mucked around a little bit more with the hasher and the function insertion logic to take into account the inlining logic brought up here, and i'm seeing some interesting things. thank you for the step-by-step explanation of what's going on.

i've managed to get the snapshots to line up _except_ for the absolute number of functions that are present in the symcache. on the surface, this seems like this is exactly what i want: duplicate function entries are being removed.

however, as you've mentioned in 4 the snapshot creator filters out all functions with an address of `u32::MAX` in the symcache's list of functions. is it possible for 3 to never be true for a valid inline function, leaving it at an `entry_pc` of `u32::MAX` and therefore filtering out an inline function from the final snapshot used for tests?

Reply:
had a chat with @loewenheim about this and opened up a PR with just the dedupe logic in #541. any future discussion about this should go into that PR, but for now this symcache conversion PR isn't going to fuss about duplicate function entries showing up while converting usyms.